File: Algorithms/Hashing_algorithms/non_repeat_char.py
# Given a string, find the first non-repeating character

def find_first_non_repeating_char(s):
    count = {}
    for char in s:
        count[char] = count.get(char, 0) + 1

    for i, char in enumerate(s):
        if count[char] == 1:
            return i

    return -1  # No non-repeating character found

s = "openai"

index = find_first_non_repeating_char(s)
if index != -1:
    print(f"First non-repeating character: {s[index]}")
else:
    print("No non-repeating character found!")

File: Algorithms/Hashing_algorithms/non_repeat_char.c
// Given a string, find the first non-repeating character

#include <stdio.h>
#include <string.h>

#define MAX_SIZE 256

int findFirstNonRepeatingChar(char* str) {
    int count[MAX_SIZE] = {0};
    int n = strlen(str);

    for (int i = 0; i < n; i++) {
        count[str[i]]++;
    }

    for (int i = 0; i < n; i++) {
        if (count[str[i]] == 1) {
            return i;
        }
    }

    return -1;  // No non-repeating character found
}

int main() {
    char str[] = "openai";

    int index = findFirstNonRepeatingChar(str);
    if (index != -1) {
        printf("First non-repeating character: %c\n", str[index]);
    } else {
        printf("No non-repeating character found!\n");
    }

    printf("\n");
    return 0;
}
